### DISCUSSION TITLE
Is it possible to update the minor version of Python <3.MAJOR.MINOR> like the windows installer?

### DISCUSSION THREAD

**Discussion by manojbaishya** [pyenv/pyenv#1985] [Q&A] [answered]
I have Python ```3.8.6``` installed with an alias ```pyenv_3.8.6``` like so:
```bash
λ VERSION_ALIAS="pyenv_3.8.6" pyenv install 3.8.6
```
```bash
λ pyenv version
pyenv_3.8.6 (set by /home/manoj/.python-version)

λ python -V
Python 3.8.6
```
Running ```pyenv install --list``` reports that the latest available ```3.8``` minor version is ```3.8.10```.

Is it possible to update my local installation from ```3.8.6``` to ```3.8.10```? I know that the windows installer does this. 

### Benefit
The ```3.8.6``` installation has a lot of libraries installed in the main directory. Updating to ```3.8.10``` will allow reusing them without having to download again in a separate installation.

<details>
  <summary>

```λ pip list```

  </summary>

```
  Package                           Version      Location
--------------------------------- ------------ ---------------------------
3to2                              1.1.1
absl-py                           0.11.0
aiodns                            2.0.0
aiohttp                           3.7.3
aiohttp-cors                      0.7.0
alabaster                         0.7.12
altgraph                          0.17
amici                             0.11.8
amply                             0.1.4
anyio                             2.2.0
anytree                           2.8.0
appdirs                           1.4.4
applib                            1.2
apptools                          4.5.0
aqtinstall                        0.9.7
argcomplete                       1.12.1
argon2-cffi                       20.1.0
arrow                             1.0.3
asciinema                         2.0.2
asgiref                           3.3.1
astroid                           2.4.1
asttokens                         2.0.4
astunparse                        1.6.3
async-generator                   1.10
async-timeout                     3.0.1
atomicwrites                      1.4.0
attrdict                          2.0.1
attrs                             19.3.0
Automat                           20.2.0
autopep8                          1.5.4
Babel                             2.8.0
backcall                          0.2.0
bandit                            1.7.0
beartype                          0.5.1
beautifulsoup4                    4.9.3
beniget                           0.3.0
binaryornot                       0.4.4
black                             20.8b1
bleach                            3.2.1
blessings                         1.7
blinker                           1.4
boltons                           20.2.1
bottle                            0.12.19
bpython                           0.21
Brotli                            1.0.9
brotlipy                          0.7.0
brython                           3.9.1
bson                              0.5.10
CacheControl                      0.12.6
cached-property                   1.5.2
cachetools                        4.1.1
cachy                             0.3.0
captum                            0.2.0
cchardet                          2.1.7
certifi                           2020.6.20
cffi                              1.14.3
cGraph                            0.1
chardet                           3.0.4
charset-normalizer                1.4.1
cheap-repr                        0.4.4
cleo                              0.8.1
cli-helpers                       2.1.0
click                             7.1.2
click-man                         0.4.1
clikit                            0.6.2
cloudpickle                       1.6.0
cmdln                             2.0.0
CodeReview                        1.0.0
colorama                          0.4.4
coloredlogs                       14.0
colour                            0.1.5
commando                          1.0.0
commonmark                        0.9.1
configobj                         5.0.6
constantly                        15.1.0
control                           0.8.3
cookiecutter                      1.7.2
crashtest                         0.3.1
crayons                           0.3.1
cryptography                      3.2
cssselect                         1.1.0
curio                             1.5
curtsies                          0.3.5
cvxopt                            1.2.6
cvxpy                             1.1.12
cwcwidth                          0.1.4
cyarray                           1.1
cyberbrain                        0.0.10
cycler                            0.10.0
Cython                            0.29.23
dask                              2021.4.1
dataclasses                       0.6
DataProfiler                      0.4.6
decorator                         4.4.2
defusedxml                        0.6.0
deprecation                       2.1.0
descartes                         1.1.0
dictquery                         0.4.0
diff-match-patch                  20200713
diffeqpy                          1.1.0
dill                              0.3.3
distlib                           0.3.1
distributed                       2021.4.1
distro                            1.5.0
Django                            3.1.7
dnspython                         1.16.0
doc2dash                          2.3.0
doc8                              0.8.1
docopt                            0.6.2
docstring-to-markdown             0.9
docutils                          0.16
dodgy                             0.2.1
doit                              0.33.1
dotty-dict                        1.3.0
dufte                             0.2.14
ecos                              2.0.7.post1
eg                                1.2.0
entrypoints                       0.3
envisage                          4.9.2
esprima                           4.0.1
et-xmlfile                        1.0.1
eventlet                          0.25.2
executing                         0.5.3
face                              20.1.1
fancycompleter                    0.9.1
fastavro                          1.4.1
feedgenerator                     1.9.1
fett                              0.3.2
ffpyplayer                        4.3.2
filelock                          3.0.12
flake8                            3.9.2
flake8-bandit                     2.1.2
flake8-docstrings                 1.6.0
flake8-polyfill                   1.0.2
Flask                             1.1.2
Flask-Caching                     1.10.1
Flask-Compress                    1.8.0
Flask-Cors                        3.0.10
Flask-SocketIO                    2.9.6
Flask-SQLAlchemy                  2.5.1
flatbuffers                       1.12
fluiddyn                          0.3.3
fluidfft                          0.3.2
fonttools                         4.18.2
fortran-language-server           1.12.0
freetype-py                       2.2.0
fs                                2.4.12
fsspec                            2021.4.0
fswrap                            1.0.0
future                            0.18.2
gast                              0.4.0
Genshi                            0.7.5
gensim                            3.8.3
get-port                          0.0.5
gevent                            1.5.0
gevent-websocket                  0.10.1
ghp-import                        1.0.0
gitdb                             4.0.5
GitPython                         3.1.11
glcontext                         2.3.3
glfw                              2.0.0
glom                              20.11.0
glumpy                            1.1.0
google-auth                       1.23.0
google-auth-oauthlib              0.4.2
google-pasta                      0.2.0
graphviz                          0.16
greenlet                          0.4.16
grid-strategy                     0.0.1
grpcio                            1.38.0
grpcio-tools                      1.36.1
gtabview                          0.8
gym                               0.17.3
h11                               0.12.0
h2                                3.2.0
h5netcdf                          0.8.1
h5py                              2.10.0
hachiko                           0.3.0
harold                            1.0.1
HeapDict                          1.0.1
helpdev                           0.7.1
highlight-text                    0.2
hpack                             3.0.0
html5lib                          1.1
httpcore                          0.12.3
httpie                            2.4.0
httpx                             0.17.1
humanfriendly                     8.2
husl                              4.0.3
hyperframe                        5.2.0
hyperlink                         21.0.0
icecream                          2.1.0
idlex                             1.18
idna                              2.10
imagesize                         1.2.0
importlib-metadata                2.0.0
incremental                       21.3.0
inflect                           5.0.2
inflection                        0.5.1
Inflector                         3.0.1
iniconfig                         1.1.1
intel-openmp                      2018.0.3
intervaltree                      3.1.0
invoke                            1.5.0
ipykernel                         5.3.4
ipython                           7.18.1
ipython-genutils                  0.2.0
ipywidgets                        7.5.1
isort                             4.3.21
isso                              0.12.5
itemadapter                       0.2.0
itemloaders                       1.0.4
itsdangerous                      1.1.0
jdcal                             1.4.1
jedi                              0.18.0
jedi-language-server              0.30.3
jeepney                           0.4.3
Jinja2                            2.11.2
jinja2-time                       0.2.0
jmespath                          0.10.0
joblib                            0.17.0
johnnydep                         1.7
json-rpc                          1.13.0
json5                             0.9.5
jsonpickle                        1.4.1
jsonschema                        3.2.0
julia                             0.5.6
jupinx                            0.2.3
jupyter                           1.0.0
jupyter-book                      0.10.2
jupyter-cache                     0.4.2
jupyter-client                    6.1.7
jupyter-console                   6.2.0
jupyter-contrib-core              0.3.3
jupyter-contrib-nbextensions      0.5.1
jupyter-core                      4.6.3
jupyter-highlight-selected-word   0.2.0
jupyter-latex-envs                1.4.6
jupyter-lsp                       0.9.2
jupyter-nbextensions-configurator 0.4.1
jupyter-packaging                 0.9.2
jupyter-server                    1.6.4
jupyter-server-mathjax            0.2.2
jupyter-sphinx                    0.3.2
jupyterbook-latex                 0.2.0
jupyterlab                        2.2.9
jupyterlab-pygments               0.1.2
jupyterlab-server                 1.2.0
jupytext                          1.10.3
kaggle                            1.5.12
Keras                             2.4.3
Keras-Preprocessing               1.1.2
keyring                           21.4.0
Kivy                              2.0.0
Kivy-examples                     2.0.0
Kivy-Garden                       0.1.4
kiwisolver                        1.2.0
language-tool-python              2.5.3
latexcodec                        2.0.1
lazy-object-proxy                 1.4.3
lets-plot                         1.5.6
libclang-py3                      3.9.0
lineedit                          0.1.5
linkify-it-py                     1.0.1
livereload                        2.6.3
llist                             0.7.1
llvmlite                          0.34.0
locket                            0.2.1
lockfile                          0.12.2
lunr                              0.5.8
lxml                              4.6.1
Mako                              1.1.3
manim                             0.5.0
manimgl                           1.0.0
ManimPango                        0.2.5.post0
mapbox-earcut                     0.12.10
Markdown                          3.2.2
markdown-it-py                    0.6.2
markdown-word-count               0.0.1
Markups                           3.0.0
MarkupSafe                        0.23
matplotlib                        3.3.2
mayavi                            4.7.2
mccabe                            0.6.1
mdit-py-plugins                   0.2.6
MechanicalSoup                    0.12.0
micawber                          0.5.2
mimesis                           4.1.3
minchin.pelican.plugins.nojekyll  1.0.2
misaka                            2.1.1
mistune                           0.8.4
mizani                            0.7.3
mkdocs                            1.1.2
mkl                               2019.0
mkl-include                       2019.0
mod-wsgi                          4.7.1
moderngl                          5.6.4
moderngl-window                   2.3.0
monotonic                         1.5
more-itertools                    8.6.0
mpi4py                            3.0.3
mpmath                            1.1.0
msgpack                           1.0.0
multidict                         5.0.2
multipledispatch                  0.6.0
multitasking                      0.0.9
mypy                              0.812
mypy-extensions                   0.4.3
mysqlclient                       2.0.2
myst-nb                           0.12.3
myst-parser                       0.13.7
natsort                           7.1.0
nbclient                          0.5.1
nbconvert                         5.6.1
nbdime                            3.0.0
nbformat                          5.0.8
nbsphinx                          0.8.4
nest-asyncio                      1.4.2
nested-lookup                     0.2.22
networkx                          2.5
NEURON                            7.8.2
Nikola                            8.1.2
nlopt                             2.6.2
nltk                              3.5
nose                              1.3.7
notebook                          6.1.4
Nuitka                            0.6.15
numba                             0.51.2
numexpr                           2.7.1
numpy                             1.20.3
numpydoc                          1.1.0
nvidia-ml-py                      11.450.51
oauthlib                          3.1.0
OMPython                          3.3.0
openpyxl                          3.0.5
opt-einsum                        3.3.0
or-datasets                       0.0.3
ortools                           8.1.8487
osqp                              0.6.1
outcome                           1.1.0
oyaml                             1.0
packaging                         20.4
paho-mqtt                         1.5.1
palettable                        3.3.0
panda3d                           1.10.9
pandarallel                       1.5.2
pandas                            1.1.3
pandas-datareader                 0.9.0
pandastable                       0.12.2.post1
pandocfilters                     1.4.3
parameterized                     0.7.4
parsel                            1.6.0
parso                             0.8.2
partd                             1.2.0
pastel                            0.2.1
path                              15.0.1
pathspec                          0.8.0
pathtools                         0.1.2
patsy                             0.5.1
pbr                               5.5.1
pdbpp                             0.10.2
pdbr                              0.5.1
pdoc                              7.0.2
pelican                           4.6.0
pelican-image-process             2.0.0
pelican-jinja2content             1.0.1
pelican-jupyter                   0.10.0
pelican-katex                     1.6.1
pelican-neighbors                 1.1.0
pelican-pandoc-reader             1.0.0
pelican-related-posts             1.0.0
pelican-render-math               1.0.3
pelican-similar-posts             1.0.0
pelican-simple-footnotes          1.0.2
pelican-sitemap                   1.0.2
pelican-thumbnailer               1.0.0
pendulum                          2.1.2
pep517                            0.9.1
pep8-naming                       0.10.0
perfplot                          0.8.10
petsc                             3.14.0
pex                               2.1.39
pexpect                           4.8.0
pgcli                             3.1.0
pgspecial                         1.11.10
phpserialize                      1.3
pickleshare                       0.7.5
PICOS                             2.0.13
piexif                            1.1.3
Pillow                            8.2.0
pip                               21.1.2
pipdeptree                        2.0.0
pipx                              0.15.5.1
pkgconfig                         1.5.1
pkginfo                           1.6.0
plotly                            4.12.0
plotnine                          0.8.0
plotnine-examples                 0.0.4
pluggy                            0.13.1
ply                               3.11
poyo                              0.5.0
pprintjson                        1.4.2
priority                          1.3.0
progressbar                       2.5
prometheus-client                 0.8.0
prompt-toolkit                    3.0.18
prospector                        1.3.1
Protego                           0.1.16
protobuf                          3.14.0
psutil                            5.7.3
psycopg2                          2.8.6
psycopg2-mq                       0.5.3
ptpython                          3.0.17
ptyprocess                        0.6.0
PuLP                              2.4
py                                1.9.0
py4j                              0.10.9
py7zr                             0.9.9
pyarrow                           3.0.0
pyasn1                            0.4.8
pyasn1-modules                    0.2.8
pyblas                            0.0.9
pybtex                            0.24.0
pybtex-docutils                   1.0.0
pybullet                          3.0.6
pycairo                           1.20.0
pycares                           3.1.1
pycodestyle                       2.7.0
pycparser                         2.20
pycryptodome                      3.9.9
pycuda                            2020.1
pydantic                          1.8.1
pydata-sphinx-theme               0.6.3
PyDispatcher                      2.0.5
pydocstyle                        6.1.1
pydot                             1.4.1
pydub                             0.25.1
pyenchant                         3.1.1
pyface                            7.1.0
pyFFTW                            0.12.0
pyflakes                          2.3.1
pygal                             2.4.0
pygame                            2.0.1
pygdbmi                           0.10.0.0
pygit2                            1.4.0
pyglet                            1.5.17
pygls                             0.10.3
Pygments                          2.9.0
PyGObject                         3.38.0
pygpu                             0.7.6
pygraphviz                        1.7
pyhull                            2015.2.1
pyinotify                         0.9.6
pyinstaller                       4.3
pyinstaller-hooks-contrib         2021.1
PyJSONViewer                      1.4.4
PyJWT                             1.7.1
pykwalify                         1.7.0
pylama                            7.7.1
pylev                             1.3.0
pyline                            0.3.20
pylint                            2.5.3
pylint-celery                     0.3
pylint-django                     2.1.0
pylint-flask                      0.6
pylint-plugin-utils               0.6
pyls                              0.1.6
pyls-black                        0.4.6
pyls-flake8                       0.3.0
pyls-isort                        0.2.0
pyls-mypy                         0.1.8
pyls-spyder                       0.3.2
pymongo                           3.11.2
PyMySQL                           0.10.1
pyodesys                          0.13.1
PyOpenGL                          3.1.5
pyOpenSSL                         20.0.1
pyparsing                         2.4.7
pyperclip                         1.8.2
Pyphen                            0.10.0
pypng                             0.0.20
pyportfolioopt                    1.4.1
PyQt5                             5.15.4
pyqt5-plugins                     5.15.4.2.1.0
PyQt5-Qt5                         5.15.2
PyQt5-sip                         12.8.1
PyQt5-stubs                       5.15.2.0
pyqt5-tools                       5.15.2.3
PyQtChart                         5.15.2
pyqtgraph                         0.11.0
PyQtWebEngine                     5.12.1
pyreadline                        2.1
pyrepl                            0.9.0
pyrr                              0.10.3
pyrsistent                        0.17.3
PyRSS2Gen                         1.1
PySDL2                            0.9.7
pyserial                          3.4
PySide2                           5.15.2
PySide6                           6.0.0
PySnooper                         0.4.2
PySocks                           1.7.1
pyspark                           3.1.1
pyspread                          1.99.5
pytest                            6.1.1
python-box                        5.2.0
python-dateutil                   2.8.1
python-dotenv                     0.14.0
python-engineio                   3.13.2
python-jsonrpc-server             0.4.0
python-language-server            0.36.2
python-libsbml                    5.18.3
python-lsp-black                  1.0.0
python-lsp-jsonrpc                1.0.0
python-lsp-server                 1.0.1
python-markdown-math              0.8
python-slugify                    4.0.1
python-snappy                     0.6.0
python-socketio                   4.6.1
python-xlib                       0.29
pythran                           0.9.10
pytomlpp                          1.0.2
pytools                           2020.4.3
pytz                              2020.1
pytzdata                          2020.1
pyuv                              1.4.0
pyxdg                             0.27
PyYAML                            5.3.1
pyzmq                             21.0.0
QDarkStyle                        3.0.2
qddate                            0.1.1
qpsolvers                         1.6.1
QScintilla                        2.12.1
qstylizer                         0.1.10
qt5-applications                  5.15.2.2.1
qt5-tools                         5.15.2.1.0.1
QtAwesome                         1.0.2
qtconsole                         5.0.3
QtPy                              1.9.0
quadprog                          0.1.8
quantecon                         0.5.0
queuelib                          1.5.0
radian                            0.5.9
raylib-py                         0.1.1
rchitect                          0.3.29
rdatasets                         0.1.0
readme-renderer                   28.0
regex                             2020.10.15
requests                          2.25.1
requests-oauthlib                 1.3.0
requests-toolbelt                 0.9.1
requirements-detector             0.7
restructuredtext-lint             1.3.2
restview                          2.9.2
retrying                          1.3.3
rfc3986                           1.4.0
rich                              10.1.0
rope                              0.18.0
rsa                               4.6
rst2html5                         2.0
rstcheck                          3.3.1
rstviewer                         0.0.6
ruamel.yaml                       0.16.12
ruamel.yaml.clib                  0.2.2
scalene                           1.3.2
scikit-cuda                       0.5.3
scikit-glpk                       0.4.1
scikit-learn                      0.23.2
scikits.odes                      2.6.1
scipy                             1.6.3
Scrapy                            2.5.0
screeninfo                        0.6.7
scs                               2.1.2
seaborn                           0.11.0
SecretStorage                     3.1.2
Send2Trash                        1.5.0
service-identity                  18.1.0
setoptconf                        0.2.0
setproctitle                      1.2.1
setuptools                        51.3.3
setuptools-scm                    6.0.1
shellingham                       1.3.2
shiboken2                         5.15.2
shiboken6                         6.0.0
shortuuid                         1.0.1
simplejson                        3.17.2
simpy                             4.0.1
six                               1.15.0
skorch                            0.9.0
slugify                           0.0.1
slycot                            0.4.0
smart-open                        4.1.0
smartypants                       1.8.6
smmap                             3.0.4
sniffio                           1.2.0
snoop                             0.2.5
snowballstemmer                   2.0.0
sortedcollections                 2.1.0
sortedcontainers                  2.2.2
soupsieve                         2.1
spatialmath-python                0.8.3
Sphinx                            4.0.2
sphinx-book-theme                 0.1.0
sphinx-comments                   0.0.3
sphinx-copybutton                 0.3.1
sphinx-glpi-theme                 0.3
sphinx-panels                     0.5.2
sphinx-thebe                      0.0.8
sphinx-togglebutton               0.2.3
sphinxcontrib-applehelp           1.0.2
sphinxcontrib-bibtex              2.1.4
sphinxcontrib-devhelp             1.0.2
sphinxcontrib-htmlhelp            1.0.3
sphinxcontrib-jsmath              1.0.1
sphinxcontrib-jupyter             0.5.10
sphinxcontrib-qthelp              1.0.3
sphinxcontrib-serializinghtml     1.1.4
spiceypy                          3.1.1
SQLAlchemy                        1.3.20
sqlparse                          0.4.1
statsmodels                       0.12.2
stevedore                         3.2.2
structlog                         21.1.0
sym                               0.3.4
sympy                             1.6.2
tablexplore                       0.1.0        /home/manoj/src/tablexplore
tabulate                          0.8.7
tavern                            1.14.2
tblib                             1.7.0
tensorboard                       2.5.0
tensorboard-data-server           0.6.1
tensorboard-plugin-wit            1.7.0
tensorflow                        2.3.1
tensorflow-estimator              2.4.0
tensorflow-gpu                    2.4.1
termcolor                         1.1.0
terminado                         0.9.1
terminaltables                    3.1.0
testpath                          0.4.4
text-unidecode                    1.3
textdistance                      4.2.0
texttable                         1.6.3
Theano                            1.0.5
thonny                            3.2.7
threadpoolctl                     2.1.0
three-merge                       0.1.1
tinycss                           0.4
tldr                              1.1.0
toml                              0.10.2
tomli                             1.0.1
tomlkit                           0.7.0
toolz                             0.11.1
torch                             1.7.0+cu110
torchaudio                        0.7.0
torchvision                       0.8.1+cu110
tornado                           6.1
tqdm                              4.51.0
traitlets                         5.0.5
traits                            6.1.1
traitsui                          7.1.0
transaction                       3.0.1
transonic                         0.4.5
triangle                          20200424
trio                              0.18.0
Twisted                           21.2.0
typed-ast                         1.4.1
typeguard                         2.12.0
typer                             0.3.2
typing-extensions                 3.7.4.3
typogrify                         2.0.7
uc-micro-py                       1.0.1
ujson                             4.0.2
undatum                           1.0.7
unicodedata2                      13.0.0.post2
Unidecode                         1.1.1
urllib3                           1.25.11
userpath                          1.4.1
validators                        0.18.2
virtualenv                        20.4.3
virtualenv-clone                  0.5.4
virtualenvwrapper                 4.8.4
visidata                          2.4
vispy                             0.6.5
vtk                               9.0.1
vulkan                            1.1.99.1
w3lib                             1.22.0
watchdog                          0.10.3
wcwidth                           0.2.5
webencodings                      0.5.1
Werkzeug                          1.0.1
wheel                             0.36.2
widgetsnbextension                3.5.1
wimpy                             0.6
wmctrl                            0.4
wrapt                             1.12.1
wurlitzer                         2.0.1
wxPython                          4.1.0
xar                               20.12.2
xlrd                              2.0.1
yapf                              0.30.0
Yapsy                             1.12.2
yarl                              1.6.3
yfinance                          0.1.59
youtube-dl                        2020.11.21.1
zict                              2.0.0
zipp                              3.4.0
zope.cachedescriptors             4.3.1
zope.interface                    5.2.0
zope.sqlalchemy                   1.3
zopfli                            0.1.7
```
</details>

**Comment by joshfriend** [comment, marked as answer]
use [`pyenv pip-migrate`](https://github.com/pyenv/pyenv-pip-migrate). I don't think an in-place migration is actually possible, because some packages have native compiled extensions?

**Reply by manojbaishya** [reply]
Sorry, I do not know if native compiled extensions depend on minor version as well.
Per [pyenv-pip-migrate](https://github.com/pyenv/pyenv-pip-migrate)'s README,
> pyenv-pip-migrate uses pip freeze to dump all installed packages in a Python version, and then tries to pip install them into another version.

My internet is poor at the moment, but it looks like ```pyenv-pip-migrate``` first writes a ```requirements.txt``` from the old version and uses ```pip install --requirement requirements.txt``` into the new version, which will redownload all packages.  I can hypothesize why this may be necessary. For some packages like numpy, I had manually compiled it with Intel oneAPI MKL by setting flags at ```~/.numpy-site.cfg``` and downloading source instead of wheel, so I guess it will not be compatible with the new version and recompilation will be required.

I think the best option will be to use ```pip install --user <pkg>``` for all future installs, which will enable sharing from ```~/.local/lib/pythonX.Y```.

In-place update would be desirable but I guess above method is okay.
